packages/flutter_tools/lib/src/runner/target_devices.dart

```diff
@@ -486,25 +503,32 @@ class TargetDevicesWithExtendedWirelessDeviceDiscovery extends TargetDevices {
       // If there are multiple matches, continue on to wait for all attached
       // and wireless devices to load so the user can select between all
       // connected matches.
-      final List<Device> devices = await _getDeviceById(
+      List<Device> specifiedDevices = await _getDeviceById(
         includeDevicesUnsupportedByProject: includeDevicesUnsupportedByProject,
         includeDisconnected: true,
       );
-      if (devices.length == 1) {
-        Device? matchedDevice = devices.first;
+
+      specifiedDevices = await _removeDevDisabledIOSDevices(specifiedDevices);
+
+      if (specifiedDevices.length == 1) {
+        Device? matchedDevice = specifiedDevices.first;
+
         if (!matchedDevice.isConnected && matchedDevice is IOSDevice) {
           matchedDevice = await _waitForIOSDeviceToConnect(matchedDevice);
         }
+
         if (matchedDevice != null && matchedDevice.isConnected) {
           return <Device>[matchedDevice];
         }
       }
     }
 
-    final List<Device> attachedDevices = await _getAttachedDevices(
+    List<Device> attachedDevices = await _getAttachedDevices(
       supportFilter: _defaultSupportFilter(includeDevicesUnsupportedByProject),
     );
 
+    attachedDevices = await _removeDevDisabledIOSDevices(attachedDevices, printWarning: false);
```

Comment:
Checking the attached devices is unnecessary. The device list returned by `_getDeviceById` above returns all devices, regardless of the connection type. `_getAttachedDevices` only returns devices that are connected and a device with the dev mode error will not be considered connected so would not appear in this list

https://github.com/flutter/flutter/blob/7961a17e60b25079ae7ba3315aeb78e9c8ce527d/packages/flutter_tools/lib/src/macos/xcdevice.dart#L525-L531

Reply:
Hmm, if I remove this line I definitely get dev mode disabled devices in the attachedDevices, which is why I added it in the first place.

Reply:
Hmm I just tried it and I don't. When it doubt, unit test it